Font-related fixes

vadz opened the pull request:
This PR completes and extends the recent changes introducing support for fractional point sizes and arbitrary numeric weights to wxFont. Here is what it does:

- [x] Avoid duplicating `{Get,Set}Weight()` in different ports.
- [x] Avoid duplicating `{Get,Set}PointSize()` in different ports.
- [x] Fix wxX11 build
- [x] Allow specifying numeric weights and floating point sizes in `wxFontInfo`
- [x] Update the font sample to show the new functionality (and allow testing it)
- [x] Get rid of hardcoded `wxDEFAULT_FONT_SIZE` in wxGTK.
- [x] Add at least trivial unit tests
- [x] Fix wxQt
- [x] Fix wxGTK1 (waste of time, but well...)
- [x] Fix wxDFB (sigh)
- [x] Fix wxMotif (arrgh)
- [ ] ~~Add CI builds for all these platforms (see #922)~~ Will be done in a separate PR
- [x] Remove Mac-specific `wxFont` ctor and `Create()` overloads taking `float pointSize`
- [x] Update all `wxNativeFontInfo::{To,From}String()` to deal with the fractional point sizes.
- [x] Allow floating point sizes in `wxXmlResourceHandlerImpl::GetFont()`
- [x] Update `wxDumpFont()`
- [x] Change `wxMarkupSpanAttributes::m_fontSize` to be `float`
- [ ] ~~Change `wxFontBase::SetPixelSize()` to use `SetFractionalPointSize()`~~ Not really clear how to do it, postponing
- [x] Use `GetFractionalPointSize()` in `wxGDIPlusFontData` and `wxD2DFontData` ctors

I'd like to merge this relatively quickly, please let me know if you have any comments. TIA!

vadz committed:
Extract helper functions for wxFontWeight <-> int conversions

Make these functions available for the upcoming reuse in wxFont.

This is almost a pure refactoring, except that assert checking for the
symbolic weight validity was corrected (it was always true before due to
wrong use of "||" instead of "&&", so split it in 2 asserts to ensure
that this doesn't happen any more).

vadz committed:
Move {Get,Set}Weight() implementations to wxFontBase

Implement these methods in the base class in terms of
{Get,Set}NumericWeight() in order to avoid having to do it for each
platform.

No real changes.

vadz committed:
Implement arbitrary font weight support for wxX11

This is mostly done to fix (half of) compilation errors after
introduction of pure virtual wxFont::{Get,Set}NumericWeight() as wxFont
in wxX11 couldn't be compiled any more as it didn't define them.

vadz committed:
Switch to using float for point size in wxNativeFontInfo

Use float as the fundamental type for the font size and implement
wxNativeFontInfo::{Set,Get}PointSize() as wrappers around the new
{Set,Get}FractionalPointSize().

Update wxNativeFontInfo for all platforms and replace the use of its
SetPointSize() method with SetFractionalPointSize() in wxFont for the
platforms already supporting fractional point sizes (don't change the
others just yet).

Note that wxNativeFontInfo::{Get,Set}PointSize() are preserved for
backwards compatibility but shouldn't be used in any code inside the
library itself any more (again, this is not the case yet, but will be
soon).

vadz committed:
Remove unnecessary wxUSE_FONTMAP guard from wxFont unit test

This test doesn't require wxFontMapper.

vadz committed:
Remove CppUnit boilerplate from wxFont unit test

No real changes, just remove the test case class which is not at all
necessary in this unit test.

vadz committed:
Add unit test for wxFont::GetPointSize() and special sizes

Verify that using wxDEFAULT works to create a font of default size, for
compatibility reasons, and that using -1 works in the same way too.

vadz committed:
Add unit test for wxFont::GetWeight() and special weight values

Verify that using wxNORMAL, wxBOLD and wxLIGHT still works, as they must
do for compatibility reasons.

vadz committed:
Add unit test for wxFont::GetStyle() and special style values

Similarly to the previous commit, verify that wxNORMAL, wxITALIC and
wxSLANT still work for compatibility (note the latter is synonymous with
wxITALIC under MSW).

vadz committed:
Add wxFont::SetFractionalPointSize()

Changing SetPointSize() argument type from int to float wasn't 100%
backwards-compatible as it notably started resulting in warnings (from
at least MSVC) about conversions from int to float in the existing code.

To avoid these warnings and for symmetry with GetFractionalPointSize(),
add SetFractionalPointSize() taking float argument and preserve the
argument of type int in SetPointSize() for compatibility.

SetPointSize() is now just a wrapper forwarding to the more general
SetFractionalPointSize().

Notice that the other ports still remain broken, this commit only
updates the currently working wxGTK, wxMac and wxMSW.

vadz committed:
Centralize backwards compatibility code in wxFont

Add wxFontBase::AccountForCompatValues() and use it in all ports instead
of redoing the same comparison with wxDEFAULT in all of them.

This is done not so much to avoid the code duplication, which was
minimal anyhow, but to make the code more clear and make it easier to
remove it from all ports at once in the bright (but remote) future when
we don't need these compatibility hacks any more.

Also document that wxDEFAULT and wxNORMAL are only handled specially in
the old-style ctor taking the individual font components and not the new
one using wxFontInfo and extend the unit test to check this.

vadz committed:
Fix wxX11 build after fractional point size changes

Define {Get,Set}FractionalPointSize() in wxX11 wxFont implementation.

Note that only Pango-based version really supports floating point sizes
as XLFD can't express them.

vadz committed:
Fix arbitrary numeric weights in the generic wxNativeFontInfo

Store the weight as int, not wxFontWeight, and rely on the same wrappers
as all the other ports in the port not having native wxNativeFontInfo.

This is a prerequisite for fixing wxDFB and other Unix ports not using
Pango.

vadz committed:
Fix wxDFB build after the font changes

Implement the new wxFont pure virtual methods in this port version of
this class.

Neither arbitrary weights nor fractional point sizes are actually
supported in this port however.

vadz committed:
Make converters between wxFontWeight and raw values public

These functions can be useful outside of src/common/fontcmn.cpp and,
potentially, even in the user code, so make them public methods of
wxFont.

No real changes, just add asserts verifying input argument value to
GetWeightClosestToNumericValue().

vadz committed:
Add numeric weights support for ports using XLFD-based fonts

Translate numeric weight to the symbolic name too instead of only doing
it in the other direction in GetNumericWeight().

vadz committed:
Remove unused variable from the font sample

No real changes.

vadz committed:
Partially revert "CMake: Fix building and running samples"

This reverts part of 573e887a4c00bd3e050d7ca27a3546467969feba, see
https://github.com/wxWidgets/wxWidgets/pull/666 because it broke the use
of the generic dialogs in the sample under MSW (they were not used any
more, even if they were supposed to) and the explanation in the original
commit message doesn't seem correct: the WXUSINGDLL check is precisely
supposed to check if we're using a static build of wxWidgets (presumably
because of DLL export complications when not using it) and seems to work
as expected.

This allows to test generic dialogs in the sample again under MSW.

vadz committed:
Exchange the font and test panels in the font sample

No real changes, but it seems to make more sense to show the font
information first and less important text sample below.

vadz committed:
Show font properties in controls in the font sample

This is done in preparation for allowing to change these control values,
although for now this is not implemented yet.

vadz committed:
Fix initial size of the font sample main window

Don't hard code main frame size in pixels, this is simple but wrong.

vadz committed:
Refactor float to int point size conversions once again

They will be also needed in wxFontInfo soon, so move them there and use
these functions from both wxFont and wxNativeFontInfo, as they can
depend on wxFontInfo but not the other way round.

No real changes.

vadz committed:
Move GetWeightClosestToNumericValue() to wxFontInfo too

This is similar to a recent commit which moved float to int point size
conversions to wxFontInfo and is done for the same reasons: wxFont and
wxNativeFontInfo can depend on wxFontInfo, but the converse is not true.

No real changes.

vadz committed:
Use wxNativeFontInfo ctor instead of manually setting its fields

This will be important when new fields are added to wxNativeFontInfo as
the ctor will be extended to initialize them, while this code wouldn't
do it.

No real changes yet.

vadz committed:
Store fractional point size in wxMSW wxNativeFontInfo

We can't losslessly recover the fractional point size from LOGFONT, so
store it in parallel.

vadz committed:
Add wxFontInfo ctor taking fractional font size in points

And implement support for it in the major ports.

Also add wxFontInfo::GetFractionalPointSize().

vadz committed:
Add wxFontInfo::Weight()

Also add not yet used wxFontInfo::GetNumericWeight().

vadz committed:
Implement support for numeric weights specified in wxFontInfo

Update the major ports to use wxFontInfo::GetNumericWeight() instead of
GetWeight().

vadz committed:
Allow changing the values of font properties in the sample

Use new wxFontInfo methods to allow testing setting fractional font
sizes and arbitrary numeric weight values in the sample.

vadz committed:
Use portable wxFont ctor instead of Mac-specific one

No real changes, just use wxFont ctor taking wxFontInfo instead of a
Mac-specific ctor.

This is more clear and will continue to work after Mac-specific ctor is
removed.

vadz committed:
Remove Mac-specific wxFont ctors and fix strikethrough support

Only accept fractional point sizes and numeric weights via wxFont ctor
using wxFontInfo and avoid having specific ctor/Create() overloads for
the different combinations of font properties: this is not portable (as
these ctors don't exist in the other ports) and unsustainable due to the
very real potential of combinatorial explosion as more properties are
added.

As a side-effect, fix support for stricken-through fonts under Mac,
which was broken, by adding the missing wxFontInfo::IsStrikethrough()
call.

vadz committed:
Add wxNativeFontInfo::SetSizeOrDefault() and use it in all ports

Reuse the same code for determining the default font size to use if it
wasn't specified in all major ports.

In particular, make wxGTK behaviour compatible with the rest and use the
normal font size in this case instead of the hardcoded value of 12pt.

vadz committed:
Get rid of wxDEFAULT_FONT_SIZE constant in wxGTK

Don't hardcode 12pt font size but use the size of the normal font
instead.

vadz committed:
Remove unnecessary wxFontRefData dtor

No real changes, just get rid of unnecessary code.

vadz committed:
Add wxFontInfo::Style() for consistency with Weight()

This will also be useful when changing the existing code to use
wxFontInfo().

vadz committed:
Change wxFontRefData ctor to take wxFontInfo in wxMSW

Apply the same change as was done for wxGTK in the previous commit to
wxMSW too, for the same reasons.

vadz committed:
Change wxFontRefData ctor to take wxFontInfo in wxGTK

wxFontInfo is the preferred way to create wxFont objects, so rewrite the
internal classes to use it directly and implement the other ctors as
wrappers for the one taking wxFontInfo.

vadz committed:
Change wxFontRefData ctor to take wxFontInfo in wxOSX too

Change wxOSX similarly to wxGTK and wxMSW in the two previous commits.

vadz committed:
Initialize wxFontRefData::m_info directly in wxOSX code

Instead of using default ctor and Init() later, just use the appropriate
ctor directly.

vadz committed:
Remove unnecessary wxFontRefData dtor in wxOSX too

Smart pointers will clean up the resources automatically, there is no
need to call Free() manually from the dtor.

vadz commented:
I think I've finally finished the API-affecting changes in the 3 major ports (which took much longer than I hoped, who would have thought...). I still hope to do the rest of the checklist above, but I'd really welcome any reviews/testing of this already. Note that you can use the updated font sample to do it quite easily but, of course, please also recompile your applications to find any compile-time breakages.

@csomor As warned, I've removed Mac-specific `wxFont` ctors, hopefully I didn't break anything.

csomor commented:
thanks a lot, sorry, I caused so much additional work with these extensions ..

catalinr commented on src/unix/fontutil.cpp:
```diff
@@ -219,9 +215,9 @@ wxFontEncoding wxNativeFontInfo::GetEncoding() const
+void wxNativeFontInfo::SetFractionalPointSize(float pointsize)
 {
-    pango_font_description_set_size( description, pointsize * PANGO_SCALE );
+    pango_font_description_set_size( description, static_cast<gint>(pointsize * PANGO_SCALE) );
```
Not tested, but would some king of _round_ be useful here instead of the cast to `gint`?

catalinr commented on include/wx/font.h:
```diff
@@ -216,16 +252,51 @@ class wxFontInfo
+    {
+        wxCHECK_MSG( pointSize == -1.0f ||
+                     (pointSize >= 0 && static_cast<float>(pointSize) < FLT_MAX),
+                     -1, "Invalid font point size" );
```
Why is this check written like this? `pointSize` is `int` but is compared to `-1.0f`. `static_cast` of `pointSize` is not needed. In case the check failed, simply `-1` would be returned, but for success there is another cast to `float`...

vadz replied:
There is definitely some confusion here but, worse, the check for `FLT_MAX` doesn't actually work neither, I don't know what was I thinking. I'll fix this, thanks.

catalinr commented on interface/wx/font.h:
```diff
@@ -322,9 +322,14 @@ class wxFontInfo
     /**
         Constructor setting the font size in points to use.
 
+        Until wxWidgets 3.1.2, the type of @a pointSize argument could only be
```
Wouldn't it be easier to read the "current behavior" first? Maybe something like

> The type of @ a pointSize argument can be either integer or fractional, starting with wxWidgets 3.1.2. Previously it could only be @c int.

vadz replied:
I've ended up with this:

        The canonical type of @a pointSize argument is @c float, however any
        other integer type, as well as @c double, is also accepted for
        compatibility.

        Notice that until wxWidgets 3.1.2, the type could only be @c int.

which, I think, is better than either version.

catalinr commented on src/common/fontcmn.cpp:
```diff
@@ -711,7 +781,7 @@ wxString wxNativeFontInfo::ToString() const
 void wxNativeFontInfo::Init()
 {
-    pointSize = 0;
+    pointSize = 0.0;
```
How about `0.0f` or `0.f` rather than `0.0`? At least to be consistent with other suffixed values.

vadz replied:
Yes, `f` was supposed to be there. Not sure how did I drop it, restored now, thanks.

MaartenBent commented on samples/dialogs/dialogs.h:
```diff
@@ -28,6 +28,12 @@ of MSW, MAC and OS2
     #define USE_WXUNIVERSAL 0
 #endif
 
+#ifdef WXUSINGDLL
```
This breaks the CMake build again because the generic dialogs are not compiled. The correct fix is to add the sources of the generic dialogs to the ` dialogs` sample. I'll make a PR for this.

MaartenBent replied:
While fixing this I noticed `"TOOLKIT=='PM'` in `dialogs.bkl`. It doesn't appear in any other file. Out of interest, what is this? And should I remove it?

vadz replied:
Yes, please remove it, it's a trace of unlamented OS/2 support (its shell was called Presentation Manager, hence PM).

vadz committed:
Deprecate old wxFont(int, int, int, int) ctor in wxQt

It might be better to just remove it completely as legacy code is
unlikely to be built with wxQt, but for now at least deprecate it to
make this consistent with the other ports.

vadz committed:
Fix indentation of wxQt wxFontRefData declaration

No real changes.

vadz committed:
Update wxFont in wxQt after recent base class interface changes

Trivially implement the new wxFont::{Set,Get}{FractionalPointSize,
NumericWeight}() methods in terms of the corresponding wxNativeFontInfo
methods.

vadz committed:
Improve mapping of font weights between Qt and wx

Support the entire [0, 99] range used by Qt instead of just some
selected values while still mapping all the predefined symbols to their
corresponding wx equivalents.

vadz committed:
Add "Apply" button to the font sample

Live changes don't work correctly in all ports, notably wxQt, which
doesn't seem to generate the events correctly for wxSpinCtrl, so add a
button to apply the changes.

This could be seen as a better UI anyhow.

vadz committed:
Remove unnecessary wxFontRefData::operator==() in wxQt

It wasn't used anywhere anyhow.

vadz committed:
Add wxFontRefData ctor from wxFontInfo to wxQt

This actually implements support for the fractional font sizes, as
previously the fractional part was lost when passing by the legacy
Create() taking only integer size.

vadz committed:
Implement wxNativeFontInfo::SetPixelSize() in wxQt

There doesn't seem to be any way to specify both font width and height
and pixels in Qt API, so just use QFont::setPixelSize() with the height.

vadz committed:
Add wxFontInfo::HasFaceName() and use it

No real changes, just simplify the code a little by adding this helper.

vadz committed:
More steps towards using wxFontInfo in all ports

Replace AccountForCompatValues() with InfoFromLegacyParams() which
directly constructs wxFontInfo from the old-style parameters, applying
all the compatibility hacks internally.

There are no real changes in this commit, just simplify the code further
and make wxFontInfo more central.

vadz committed:
Fix bug with point sizes in wxX11 without Pango

XLFD strings specify size in tens of points, so add the missing factor
of 10.

vadz committed:
Allow leaving the point size unspecified for ports using XLFD

Rather than defaulting to the normal font size, just leave the font size
unspecified as it might result in better font size being chosen.

vadz committed:
Remove unnecessary line from the font sample

This was mistakenly left over in the commit adding private fonts support
(547e40b114f44c528107afd28135fcc964c45619).

vadz committed:
Disable wxGraphicsContext in wxGTK1 build, it doesn't compile

There is no chance of wxGraphicsContext ever being supported in this
legacy port, but at least disable it automatically instead of requiring
to specify --disable-graphics-ctx on the command line.

vadz committed:
Disable tons of warnings given during wxGTK1 build

These warnings are not going to be fixed, so suppress them to at least
see the other ones more clearly.

vadz committed:
Don't assert in wxNativeFontInfo::SetStyle() in wxGTK1

Add missing "break" statement.

vadz committed:
Fix wxGTK1 build after wxFont API changes

Implement the new {Get,Set}{FractionalPointSize,NumericWeight} methods.

Also change wxLoadQueryFont() to use wxNativeFontInfo methods as a side
effect, to reduce code duplication and reuse the existing support for
numeric weights and fractional point sizes in wxNativeFontInfo.

vadz committed:
Fix wxMotif build after wxFont API changes

Run-time seems to be totally broken, but fonts are probably not the most
important problem in this port, so don't bother fixing it.

vadz committed:
Show fractional point size and numeric weight in wxDumpFont()

Update to show the new wxFont fields.

vadz commented:
I've fixed wxGTK1 and wxMotif compilation but the latter is completely unusable and the former very buggy. I'm not going to look into this further, but I really wonder how can anybody use either of these ports...

vadz committed:
Use SetFractionalPointSize() in wxMarkupParser code

Now that fractional sizes are supported, use them instead of rounding up
to integer size.

vadz committed:
Add support for strike through attribute to markup spans

It can be handled in exactly the same way as underlining, so just do it
instead of leaving a "TODO" comment about it.

vadz committed:
Use SetFractionalPointSize() in wxFont::Scale()

This will result in more precise scaling.

vadz committed:
Use fractional point sizes in SetWindowVariant()

Don't round the scaled font size now that we can just directly use the
fractional value.

vadz committed:
Use fractional point size when scaling fonts in wxTipDialog

Don't unnecessarily round the result to int any longer.

vadz committed:
Don't round the larger or smaller font sizes unnecessarily

Use SetFractionalPointSize() instead of SetPointSize() to avoid losing
the fractional part of the font size, if any.

vadz committed:
Use fractional point sizes in wxRichText code

When scaling, just use the floating point result directly instead of
rounding it.

vadz committed:
Update wxNativeFontInfo::{To,From}String() after wxFont changes

Handle fractional point sizes and numeric weights in the custom string
formats in the ports using those.

vadz committed:
Extract ConvertFromLegacyWeightIfNecessary() function

Make it possible to reuse just this part of GetNumericWeightOf() in the
upcoming commit.

vadz committed:
Support fractional font sizes and numeric weights in XRC

Change the code to handle them, the XRC sample to test them, the schema
to accept them and the documentation to describe them.

vadz committed:
Handle fractional point sizes in DirectWrite wxGraphicsFont

Also stop scaling the font size by DPI manually, we don't need to do it
if we use points instead of pixels. This, in turn, made it unnecessary
to pass ID2D1Factory parameter to wxD2DFontData ctor.

vadz committed:
Handle fractional point sizes in GDI+ wxGraphicsFont

Also switch to using UnitPoint when creating GDI+ fonts as it's not
clear at all why should we use UnitPixel here.

vadz committed:
Add strikethrough support for fonts defined in XRC

Handle this attribute as well as "underlined" for completeness.

MaartenBent commented on interface/wx/font.h:
```diff
@@ -1059,19 +1117,30 @@ class wxFont : public wxGDIObject
+        wxWidgets 3.1.2) SetFractionalPointSize() function instead.
+     */
+
+    virtual void SetPointSize(float pointSize);
```
This should be `int`?

vadz replied:
Yes, thank you.

MaartenBent reviewed (commented):
Maybe it is useful to add tests for `[Get,Set]NumericWeight` and `[Get,Set]FractionalPointSize`.

I did not notice issues with MSW port.

vadz committed:
Add simple unit tests for the new wxFont features

Check that calling SetFractionalPointSize() and SetNumericWeight() at
the very least results in the expected return values from
GetFractionalPointSize() and GetNumericWeight().

vadz committed:
Use only wxFont ctor from wxFontInfo in the samples

Follow the documentation advice and prefer to use the ctor taking
wxFontInfo as it makes the code slightly shorter and more readable.
